### ISSUE TITLE
Picking unused account in estimator more efficiently

### ISSUE THREAD

**Issue by jakmeier** [near/nearcore#7450]
## Task description
Improve the efficiency / algorithmic complexity of finding a random unused account in the runtime parameter estimator's transaction builder.

## Currently inefficient code
`random_unused_account` and `random_account_pair` in https://github.com/near/nearcore/blob/master/runtime/runtime-params-estimator/src/transaction_builder.rs both use a trial-and-error loop to find an account that is not already in use.

```
pub(crate) fn random_unused_account(&mut self) -> AccountId {
        loop {
            let account = self.random_account();
            if self.used_accounts.insert(account.clone()) {
                return account;
            }
        }
    }
```

This can be inefficient, especially when there are only few accounts left.

## Possible solutions
One way to resolve it would be keep a hash set of used accounts instead of one for unused accounts. Then a random key from the hash set could be chosen. But I haven't thought through the details if it really works. And there are probably other viable solutions.

**Comment by DevSabb** [comment]
I have a suggestion for `random_unused_account`. How about do a [random shuffle](https://rust-random.github.io/rand/rand/seq/trait.SliceRandom.html#tymethod.shuffle) (O(n) apparently) in the `new` method and pop items from the list whenever we need a new unused account? Let me know if it works, I'll take a shot at it.

**Comment by jakmeier** [comment]
So would you then keep two vectors of accounts, one for used accounts and one for unused?

**Comment by DevSabb** [comment]
@jakmeier yes. `accounts` will remain as is. `used_accounts` will be replaced with a shuffled Vector. It looks like [removing last element](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.pop) from Vector is O(1) operation.

**Comment by jakmeier** [comment]
@DevSabb That should work! Either strings or integers are fine.

The option with integer indices reduces the memory footprint a bit. One 8-byte integer per account amounts to 8 MB per 1 million accounts. Whereas a string is a fat pointer (16 bytes I think) plus the bytes of the string itself. Probably that would be about 4 times more than integers.

Although we are sometimes running this with millions of accounts, there is no good reason to do so. We only need a few thousand active accounts to select from. Therefore, the memory footprint should be fine either way. I will leave it up to you whether you want to do that optimization or not.

Thanks for looking into it!

**Comment by jakmeier** [comment]
Resolved by #7493
